src/gui/speedlimitdialog.cpp

```diff
@@ -50,57 +52,116 @@ SpeedLimitDialog::~SpeedLimitDialog()
     delete m_ui;
 }
 
-// -2: if cancel
-long SpeedLimitDialog::askSpeedLimit(QWidget *parent, bool *ok, const QString &title, const long defaultVal, const long maxVal)
+bool SpeedLimitDialog::askNewSpeedLimits(QWidget *parent, const QString &title, SpeedLimits &speedLimits)
 {
-    if (ok) *ok = false;
-
     SpeedLimitDialog dlg(parent);
     dlg.setWindowTitle(title);
-    dlg.setupDialog((maxVal / 1024.), (defaultVal / 1024.));
+    dlg.setupDialog(speedLimits);
 
-    if (dlg.exec() == QDialog::Accepted) {
-        if (ok) *ok = true;
+    if (dlg.exec() != QDialog::Accepted)
+        return false;
 
-        const int val = dlg.getSpeedLimit();
-        if (val < 0)
-            return 0;
-        return (val * 1024);
-    }
+    if (dlg.getUploadSpeedLimit() < 0)
+        speedLimits.uploadLimit = 0;
+    else
+        speedLimits.uploadLimit = (dlg.getUploadSpeedLimit() * 1024);
 
-    return -2;
+    if (dlg.getDownloadSpeedLimit() < 0)
+        speedLimits.downloadLimit = 0;
+    else
+        speedLimits.downloadLimit = (dlg.getDownloadSpeedLimit() * 1024);
+
+    return true;
+}
+
+void SpeedLimitDialog::hideShowWarning()
```

Comment:
Show or hide a warning if the limits chosen in the dialog exceed the global limits (if any). Because you know, they won't actually exceed the global limits (libtorrent hardcoded).

Reply:
Sorry, I guess I don't know about libtorrent hardcoded limits. What do you mean. Can you provide me link to the corresponding docs?

Reply:
Oh sorry i thought you knew because IIRC you said it to someone else a few years ago.
`torrent_handle::set_download_limit`
https://www.libtorrent.org/reference-Core.html (no direct link)
>set_upload_limit will limit the upload bandwidth used by this particular torrent to the limit you set. It is given as the number of bytes per second the torrent is allowed to upload. set_download_limit works the same way but for download bandwidth instead of upload bandwidth. Note that setting a higher limit on a torrent then the global limit (settings_pack::upload_rate_limit) will not override the global rate limit. The torrent can never upload more than the global rate limit.

Also applies to download limit just checked, even though the docs only mention upload limit explicitly.

Reply:
Ok, thanks.
But I was confused of "hardcoded limits" you talked about. There are no such "hardcoded" limits but just configured global limits.

Reply:
>Show or hide a warning if the limits chosen in the dialog exceed the global limits 

Aren't you putting too much emphasis on that?
I don't think this is such a strong issue that it requires this extra logic. This may be quite a normal case when per-torrent limits exceed the global ones (e.g. when alt limits are used now but user want to limit some torrent to not exceed main global limit).
I would make this dialog to be more "static", i.e. just add some (always visible) "note" about this behavior:
```
Note that setting a higher limit on a torrent then the global limit will not override the global speed limit. The torrent can never upload/download more than the global speed limit.
```
Then you will get a simple dialog that allows you to set (edit) any speed limit without any (doubtful, IMO) extra logic.

```c++
bool SpeedLimitDialog::askNewSpeedLimits(QWidget *parent, const QString &title, SpeedLimit &limit);
```

Reply:
I thought it'd be a nice touch but ok i'll make it a static note.

Reply:
BTW, one common comment. Using word "New" in such names like `SpeedLimitDialog::askNewSpeedLimits()` is (at least) redundant in most cases.

Reply:
I like explicit but i'll undo it then.

Reply:
@glassez, @thalieht:  Just a slight wording correction… 

> Note that setting a higher limit on a torrent *then* the global limit will not override the global speed limit. The torrent can never upload/download *more* than the global speed limit. 

…should be… 

> Note that setting a higher limit on a torrent *than* the global limit will not override the global speed limit. The torrent can never upload/download *faster* than the global speed limit.

Reply:
@LordNyriox that quote is from the libtorrent docs.

Reply:
then => than :+1: